### PR TITLE
[promise.allsettled] Added option to provide a type for PromiseRejection when using the allSettled function

### PR DESCRIPTION
**Pull request by nikmash**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by nikmash** [comment]
Was seeing compiler errors when trying to use the `reason` property returned by the `allSettled` function because the type is `unknown` in an error scenario. The underlying type takes in a generic so if we are able to use a generic in the `allSettled` function then the compiler will be aware of the type that the `reason` property has in the `PromiseRejection` scenario.

**Comment by nikmash** [comment]
@ljharb @jurca if y'all don't mind please review! 🙏

File: types/promise.allsettled/implementation.d.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ type PromiseResultTuple<T extends [unknown, ...unknown[]]> = {[P in keyof T]: Pr
 
 declare function allSettled(): Promise<[]>;
 declare function allSettled<T extends [unknown, ...unknown[]]>(iterable: PromiseTuple<T>): Promise<PromiseResultTuple<T>>;
-declare function allSettled<T>(iterable: Iterable<Promise<T> | T>): Promise<Array<PromiseResult<T>>>;
+// tslint:disable-next-line no-unnecessary-generics
+declare function allSettled<T, E>(iterable: Iterable<Promise<T> | T>): Promise<Array<PromiseResult<T, E>>>;
```
Comment (ljharb) [inline types/promise.allsettled/implementation.d.ts]: should E default to unknown, so it’s optional?

**Comment by jurca** [comment]
@nikmash Hello, and you for submitting this PR.

While I have originally toyed with allowing this (that's why it is there for the underlying type), in the end I decided against it.

The reason for that is that you cannot have a reasonable guarantee what value will be used to reject the promise, since any value, including `undefined` can be used to reject a promise or passed to the `throw` statement.

I'm pretty sure that's also the reason why TypeScript does not allow you to track the type of the value rejecting a Promise - typescript/lib/lib.es5.d.ts:1399-1414:

```typescript
interface Promise<T> {
    /**
     * Attaches callbacks for the resolution and/or rejection of the Promise.
     * @param onfulfilled The callback to execute when the Promise is resolved.
     * @param onrejected The callback to execute when the Promise is rejected.
     * @returns A Promise for the completion of which ever callback is executed.
     */
    then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): Promise<TResult1 | TResult2>;

    /**
     * Attaches a callback for only the rejection of the Promise.
     * @param onrejected The callback to execute when the Promise is rejected.
     * @returns A Promise for the completion of the callback.
     */
    catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | undefined | null): Promise<T | TResult>;
}
```

Notice that the `reason` argument passed to the `onrejected` callback is of the `any` type.

Because of this, adding this type parameter would result in a silent type cast in a disguise. That's also the reason why I opted for `unknown` instead of `any` - sacrificing some convenience for better type safety.

I hope you understand.

I am inclined to reject this PR for these reasons. Should you prefer to change the type from `unknown` to `any` to match the declarations that are included with TypeScript, I would be willing to accept that.

Do you agree, @ljharb ?

**Comment by ljharb** [comment]
I don’t agree; any value can be in almost any position, but the point of TS is to narrow that down, and TS *should* allow the type of a rejected promise to be similarly restricted. (you make a good case that until TS fixes this, that these types probably shouldn’t attempt to)

unknown is certainly better than any, always - but what’s the point of a type system if it’s not going to be used to narrow types at every opportunity?

**Comment by jurca** [comment]
Well, I would prefer keeping the `unknown` type as well - that's why I used it in the first place.

**Comment by nikmash** [comment]
@jurca in this case I'm using the `Axios` module which would actually have a type of `AxiosError` in the case that the promise is rejected. That's why I need to define a type for `E`.

**Comment by jurca** [comment]
@nikmash I understand, however, you have no guarantee that the error will always be an `AxiosError` instance.

I would recommend to use the following code in your case:

```typescript
const results = await allSettled([...])
for (const result of results) {
  if (result.status === 'rejected') {
    if (result.reason instanceof AxiosError) {
     const reason = result.reason as AxiosError
     // do something
    } else {
      throw result.reason
    }
  }
}
```

**Comment by nikmash** [comment]
@jurca ah I had to dig through the typescript code, but now I understand what you mean by the fact that there's no guarantee that the error is an `AxiosError`. Although the code above wouldn't work because `AxiosError` is an interface not a value so you cannot call `instanceof` on an interface. 

Instead of `result.reason instanceof AxiosError` I believe a User Defined type guard would need to be used like this:
```
  isAxiosError(arg: any): arg is AxiosError {
    return arg && arg.isAxiosError && typeof(arg.isAxiosError) === "boolean";
  }
```

But thanks for your help @jurca I think this makes sense based on the typescript spec. Until typescript narrows down the possible types of a rejected promise I guess we can't fix this.

**Pull request closed without merging**